src/wp-includes/class-wp-token-map.php

```diff
@@ -0,0 +1,810 @@
+	 */
+	public function contains( $word, $case_sensitivity = 'case-sensitive' ) {
+		$ignore_case = 'case-insensitive' === $case_sensitivity;
+
+		if ( $this->key_length >= strlen( $word ) ) {
+			if ( 0 === strlen( $this->small_words ) ) {
+				return false;
+			}
+
+			$term    = str_pad( $word, $this->key_length + 1, "\x00", STR_PAD_RIGHT );
+			$word_at = $ignore_case ? stripos( $this->small_words, $term ) : strpos( $this->small_words, $term );
+			if ( false === $word_at ) {
+				return false;
+			}
+
+			return true;
+		}
+
+		$group_key = substr( $word, 0, $this->key_length );
+		$group_at  = $ignore_case ? stripos( $this->groups, $group_key ) : strpos( $this->groups, $group_key );
+		if ( false === $group_at ) {
+			return false;
+		}
+		$group        = $this->large_words[ $group_at / ( $this->key_length + 1 ) ];
+		$group_length = strlen( $group );
+		$slug         = substr( $word, $this->key_length );
+		$length       = strlen( $slug );
+		$at           = 0;
+
+		while ( $at < $group_length ) {
+			$token_length   = unpack( 'C', $group[ $at++ ] )[1];
+			$token_at       = $at;
+			$at            += $token_length;
+			$mapping_length = unpack( 'C', $group[ $at++ ] )[1];
+			$mapping_at     = $at;
+
+			if ( $token_length === $length && 0 === substr_compare( $group, $slug, $token_at, $token_length, $ignore_case ) ) {
+				return true;
+			}
+
+			$at = $mapping_at + $mapping_length;
+		}
+
+		return false;
+	}
+
+	/**
+	 * If the text starting at a given offset is a lookup key in the map,
+	 * return the corresponding transformation from the map, else `false`.
+	 *
+	 * This function returns the translated string, but accepts an optional
+	 * parameter `$skip_bytes` which communicates how many bytes long the
+	 * lookup key was, if it found one. This can be used to advance a cursor
+	 * in calling code if a lookup key was found.
+	 *
+	 * Example:
+	 *
+	 *     false === $smilies->read_token( 'Not sure :?.', 0, $bytes_skipped );
+	 *     '😕'  === $smilies->read_token( 'Not sure :?.', 9, $bytes_skipped );
+	 *     2     === $bytes_skipped;
+	 *
+	 * Example:
+	 *
+	 *     while ( $at < strlen( $input ) ) {
+	 *         $next_at = strpos( $input, ':', $at );
+	 *         if ( false === $next_at ) {
+	 *             break;
+	 *         }
+	 *
+	 *         $smily = $smilies->read_token( $input, $next_at, $bytes_skipped );
+	 *         if ( false === $next_at ) {
+	 *             ++$at;
+	 *             continue;
+	 *         }
+	 *
+	 *         $prefix  = substr( $input, $at, $next_at - $at );
+	 *         $at     += $bytes_skipped;
+	 *         $output .= "{$prefix}{$smily}";
+	 *     }
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param string  $text              String in which to search for a lookup key.
+	 * @param ?int    $offset            How many bytes into the string where the lookup key ought to start.
+	 * @param ?int    &$skip_bytes       Holds byte-length of found lookup key if matched, otherwise not set.
```

Comment:
This parameter is very interesting. It seems very important, it's a bit like `$matches` in `preg_match` in that it's an output of the function and not an input.

I think the name `skip_bytes` can be improved. That's very focused on what calling code would likelye do with the value and less what the value _is_. Maybe something like `$matched_token_bytelength` or would be a descriptive name of what it is?

Reply:
Yes, please. It was nearly impossible for me to understand the exact purpose of this `$skip_bytes` param by looking at code examples. I'm still not sure why it would be useful for developers.

Reply:
The idea is that you're working with an input like this and using a token map to detect (and perform) replacements:

```html
true &not; false
```

We'd like to decode this by replacing entities, so we start by checking matches (this would be in a loop):

```php
$m = WP_Token_Map::from_array([ '&not;' => '¬' ]);
$s = 'true &not; false';

//  check here
//  |
//  v
// "true &not; false"
$replacement = $m->read_token( $s, 0, $skip_bytes );
// $replacement = false
// $skip_bytes = null

//   check here
//   |
//   v
// "true &not; false"
$replacement = $m->read_token( $s, 1, $skip_bytes );
// $replacement = false
// $skip_bytes = null

// …

//       check here
//       |
//       v
// "true &not; false"
$replacement = $m->read_token( $s, 5, $skip_bytes );
// $replacement = "¬"
// $skip_bytes = 5
```

At this point, we've hit a match and would perform replacement, but the function just returned `"&"` which means "at index 4 there's a token matching `&`" but we don't know what the token is so it's unclear how we should perform replacement. That's where `$skip_bytes` is involved, it lets us know the byte length of the matched token so we can replace:

```php
// 5 here is the index in the string where we found the token, we know that from our
// original iteration
substr( $s, 0, 5 ) . $replacement . substr( $s, 5 + $skip_bytes ); // "true ¬ false"
```

Reply:
Ok, so it’s the length of the token matched that we will ignore while constructing the replaced version. The coincidence that the offset and skip bytes were both equal to 5 made this puzzle more challenging, but I hope I got it right. I was mostly making the case that it was the only difficult part to reason in the PHPDoc.

Reply:
sounds like I should pull in some of the wording from the dev note and move it into the docblocks…

https://make.wordpress.org/core/?p=113042&preview=1&_ppp=452181011b

Reply:
> it’s the length of the token matched

Yep, in this case it's `strlen( "&not;" );`

> the coincidence that the offset and skip bytes were both equal to 5 made this puzzle more challenging.

🤦 yes, poor execution in my example 😓

Reply:
I've renamed the argument to `$matched_token_byte_length`. It's verbose, but more explanatory.